### ISSUE TITLE
Function interface doesn't type check return value for extra keys

### ISSUE THREAD

**Issue by carloslfu** [microsoft/TypeScript#12632]
**TypeScript Version:**  2.1.1 / nightly (2.2.0-dev.201xxxxx)

See next code, also in [This playground](https://www.typescriptlang.org/play/#src=%0D%0Ainterface%20F%20%7B%0D%0A%20%20%20%20(p%3A%20%7Ba%3Anumber%7D)%3A%20%7Ba%3A%20string%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20there%20are%20no%20type%20check%20for%20extra%20keys%0D%0A%2F%2F%20but%20F%20interface%20has%20return%20type%0D%0Alet%20g%3A%20F%20%3D%20(%7Ba%7D)%20%3D%3E%20(%7B%0D%0A%20%20%20%20a%3A%20'5'%2C%0D%0A%20%20%20%20b%3A%20''%2C%0D%0A%7D)%0D%0A%0D%0A%2F%2F%20we%20need%20to%20do%20that!%2C%20but%20F%20interface%20also%20has%20return%20type%0D%0Alet%20f%3A%20F%20%3D%20(%7Ba%7D)%3A%20%7Ba%3A%20string%7D%20%3D%3E%20(%7B%0D%0A%20%20%20%20a%3A%20'5'%2C%0D%0A%20%20%20%20b%3A%20''%2C%0D%0A%7D)%0D%0A).

**Code**

```ts
interface F {
    (p: {a:number}): {a: string}
}

// there are no type check for extra keys
// but F interface has return type
let g: F = ({a}) => ({
    a: '5',
    b: '',
})

// we need to do that!, but F interface also has return type
let f: F = ({a}): {a: string} => ({
    a: '5',
    b: '',
})
```

**Expected behavior:**

'g' function should be detected as an error, because return type of interface F doesn't have key 'b', just haves 'a' key

**Actual behavior:**

'g' function doesn't display any error

**Comment by DanielRosenwasser** [comment]
I'm not necessarily sure at which point in the process the object literal loses its [freshness](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3115-excess-properties).

@carloslfu is there any reason you're trying to prohibit excess properties here?

**Comment by carloslfu** [comment]
Suppose this scenario:

[playground link](https://www.typescriptlang.org/play/#src=%0D%0Afunction%20fun1%20(params%3A%20%7B%20a%3A%20number%2C%20b%3A%20string%20%7D)%20%3A%20%7B%20c%3A%20number%2C%20d%3A%20string%20%7D%20%7B%0D%0A%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20c%3A%200%2C%0D%0A%20%20%20%20%20%20%20%20d%3A%20''%2C%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20this%20is%20detected%20by%20compiler%0D%0Afunction%20fun2%20(params%3A%20%7B%20a%3A%20number%2C%20b%3A%20string%20%7D)%20%3A%20%7B%20c%3A%20number%2C%20d%3A%20string%20%7D%20%7B%0D%0A%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20c%3A%200%2C%0D%0A%20%20%20%20%20%20%20%20d%3A%20''%2C%0D%0A%20%20%20%20%20%20%20%20x%3A%200%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20I%20want%20to%20be%20able%20to%20do%20(to%20be%20DRY)%3A%0D%0A%0D%0Ainterface%20F%20%7B%0D%0A%20%20%20%20(params%3A%20%7B%20a%3A%20number%2C%20b%3A%20string%20%7D)%3A%20%7B%20c%3A%20number%2C%20d%3A%20string%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20fun2_v1%3A%20F%20%3D%20function%20(params)%20%7B%0D%0A%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20c%3A%200%2C%0D%0A%20%20%20%20%20%20%20%20d%3A%20''%2C%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20this%20function%20has%20the%20same%20type%20as%20fun2%2C%20but%20is%20not%20detected%20by%20compiler%0D%0Aconst%20fun2_v2%3A%20F%20%3D%20function%20(params)%20%7B%0D%0A%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20c%3A%200%2C%0D%0A%20%20%20%20%20%20%20%20d%3A%20''%2C%0D%0A%20%20%20%20%20%20%20%20x%3A%200%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A)
```ts
function fun1 (params: { a: number, b: string }) : { c: number, d: string } {
    return {
        c: 0,
        d: '',
    }
}

// this is detected by compiler
function fun2 (params: { a: number, b: string }) : { c: number, d: string } {
    return {
        c: 0,
        d: '',
        x: 0
    }
}

// I want to be able to do (to be DRY):

interface F {
    (params: { a: number, b: string }): { c: number, d: string }
}

const fun2_v1: F = function (params) {
    return {
        c: 0,
        d: '',
    }
}

// this function has the same type as fun2, but is not detected by compiler
const fun2_v2: F = function (params) {
    return {
        c: 0,
        d: '',
        x: 0
    }
}
```

**Comment by mquandalle** [comment]
As @DanielRosenwasser asked

> is there any reason you're trying to prohibit excess properties here?

I'd like give an additional use case for this feature. 

I've an API that is similar to [webpack-block](https://github.com/andywer/webpack-blocks). So the basic idea is that we want to create a configuration object with a lot of fields, and we'd like to use some form of composition to encapsulate similar patterns in that object.

To keep it simple let's say that to produce the following configuration object:

```ts
const config = {
  entry: "./main.js",
  output: {
    path: __dirname + "/build",
    filename: "bundle.js"
  }
};
```

we want to write the following code:

```ts
const config = createConfig([
  entryPoint('./src/main.js'),
  setOutput('./build/bundle.js'),
]);
```

and have the type checker ensure that the config is valid:

```ts
interface IConfig {
  entry: string;
  output: {
    path: string,
    filename: string,
  };
}
```

This composition design could almost already work. We could indeed create a partial config type:

```ts
type PartialConfig = { [P in keyof IConfig]?: IConfig[P] };
```

and have `entryPoint` and `setOuput` utilities return a `PartialConfig`, and `createConfig` take a `Array<PartialConfig>` as an input. Then if there is a typo and `setOutput` define a `fileName` instead of `filename` the compiler will catch that.

Now the problem is that we don't want `entryPoint` and `setOutput` utilities to return `PartialConfig` but instead to return a partial config transformer, ie a function of type:

```ts
type PartialConfigTransformer = (input: PartialConfig | {}) => PartialConfig;
```

that way utilities functions could execute more complex transformations on the configuration object, and not be limited to inserting additional keys.

However now if we introduce a typo again (`setOutput` uses `fileName` instead of `filename`), it won't be caught by the compiler anymore.

The minimal problem is exposed in this snippet:

![minimal example callback weak types](https://user-images.githubusercontent.com/1730702/27611135-508e0ffa-5b91-11e7-886d-ed603f16111a.png)

**Comment by andy-ms** [comment]
Just ran into this:
```ts
interface I { x: number; y: number; }
[1,2,3].map<I>(n => ({ x: n, y: n, n }));
```
This is a pretty common pattern in TypeScript's own codebase so we should support getting the contextual type on the object literal and checking for extra properties there.

**Comment by RyanCavanaugh** [comment]
@sandersn is this fixed?

**Comment by WhaleMade** [comment]
So.... any timeline on when this will be fixed?

**TypeScript Version**: 2.7.2

our use case:
```typescript
enum EConstants {
    FOO = 'foo',
    BAR = 'bar',
}

interface IState {
    id: string,
    name: string,
}

function converter<EKeys extends string, T>(
    functionMap: {[Key in EKeys] : (objectIn: T) => T}
) {
    return functionMap
}

let result = converter<EConstants, IState>({
    [EConstants.FOO]: (objectIn) => ({
        ...objectIn,

        extraProp: 'extra' // <-- this isn't detected as invalid!
                           // even though the intellisense for converter 
                           // can see that this function should return an IState type
    }),
    [EConstants.BAR]: (objectIn) => ({
        ...objectIn
    })
})
```
[Playground Link](http://www.typescriptlang.org/play/#src=enum%20EConstants%20%7B%0D%0A%20%20%20%20FOO%20%3D%20'foo'%2C%0D%0A%20%20%20%20BAR%20%3D%20'bar'%2C%0D%0A%7D%0D%0A%0D%0Ainterface%20IState%20%7B%0D%0A%20%20%20%20id%3A%20string%2C%0D%0A%20%20%20%20name%3A%20string%2C%0D%0A%7D%0D%0A%0D%0Afunction%20converter%3CEKeys%20extends%20string%2C%20T%3E(%0D%0A%20%20%20%20functionMap%3A%20%7B%5BKey%20in%20EKeys%5D%20%3A%20(objectIn%3A%20T)%20%3D%3E%20T%7D%0D%0A)%20%7B%0D%0A%20%20%20%20return%20functionMap%0D%0A%7D%0D%0A%0D%0Alet%20result%20%3D%20converter%3CEConstants%2C%20IState%3E(%7B%0D%0A%20%20%20%20%5BEConstants.FOO%5D%3A%20(objectIn)%20%3D%3E%20(%7B%0D%0A%20%20%20%20%20%20%20%20...objectIn%2C%0D%0A%0D%0A%20%20%20%20%20%20%20%20extraProp%3A%20'extra'%20%2F%2F%20%3C--%20this%20isn't%20detected%20as%20invalid!%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20even%20though%20the%20intellisense%20for%20converter%20%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20can%20see%20that%20this%20function%20should%20return%20an%20IState%20type%0D%0A%20%20%20%20%7D)%2C%0D%0A%20%20%20%20%5BEConstants.BAR%5D%3A%20(objectIn)%20%3D%3E%20(%7B%0D%0A%20%20%20%20%20%20%20%20...objectIn%0D%0A%20%20%20%20%7D)%0D%0A%7D))

**Comment by sandersn** [comment]
Object literals lose their freshness when widened, and we widen return types. I tried a couple of experiments:

1. Retain flags when widening. The compiler failed to build with this setting. Numerous casts are no longer allowed.
2. Retain flags when widening return types only. This caused freshness to persist on the function return type, which caused a couple of bugs like:

```ts
function f(name: string, age: number) { return { name, age } }
let justName: { name: string } = f("Bob", 42)
```

Now you get an error that `age` is an excess property because the fresh object literal `{ name, age }` gets persisted as the return type of `f` and of course `{ name, age }` is excess when assigned to just `{ name: string }`.

To make this work, we would need to think about all the places where freshness would need to be removed: function declarations would be one, assignments another. I think there are others.

**Comment by RyanCavanaugh** [comment]
Going to track at #241
